dataflow/run-inference/main.py

```diff
@@ -0,0 +1,142 @@
+        tokenizer: Tokenizer for the LLM model.
+
+    Returns: Tokenized input tokens.
+    """
+    return tokenizer(input_text, return_tensors="pt").input_ids[0]
+
+
+def get_response(result: PredictionResult, tokenizer: PreTrainedTokenizer) -> str:
+    """Decodes output token tensors into text.
+
+    Args:
+        result: Prediction results from the RunInference transform.
+        tokenizer: Tokenizer for the LLM model.
+
+    Returns: The model's response as text.
+    """
+    output_tokens = result.inference
+    return tokenizer.decode(output_tokens, skip_special_tokens=True)
+
+
+class AskModel(beam.PTransform):
+    """Asks an LLM a prompt message and gets its responses.
+
+    Attributes:
+        model_name: HuggingFace model name compatible with AutoModelForSeq2SeqLM.
+        state_dict_path: File path to the model's state_dict, can be in Cloud Storage.
+        max_response_tokens: Maximum number of tokens for the model to generate.
+    """
+
+    def __init__(
+        self,
+        model_name: str,
+        state_dict_path: str,
+        max_response_tokens: int = MAX_RESPONSE_TOKENS,
+    ) -> None:
+        self.model_handler = PytorchModelHandlerTensor(
+            state_dict_path=state_dict_path,
+            model_class=AutoModelForSeq2SeqLM.from_config,
+            model_params={"config": AutoConfig.from_pretrained(model_name)},
+            inference_fn=make_tensor_model_fn("generate"),
+        )
+        self.tokenizer = AutoTokenizer.from_pretrained(model_name)
+        self.max_response_tokens = max_response_tokens
+
+    def expand(self, pcollection: beam.PCollection[str]) -> beam.PCollection[str]:
+        return (
+            pcollection
+            | "To tensors" >> beam.Map(to_tensors, self.tokenizer)
+            | "RunInference"
+            >> RunInference(
+                self.model_handler,
+                inference_args={"max_new_tokens": self.max_response_tokens},
+            )
+            | "Get response" >> beam.Map(get_response, self.tokenizer)
+        )
+
+
+if __name__ == "__main__":
+    import argparse
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--messages-topic",
+        required=True,
+        help="Pub/Sub topic for input text messages",
+    )
+    parser.add_argument(
+        "--responses-topic",
+        required=True,
+        help="Pub/Sub topic for output text responses",
+    )
+    parser.add_argument(
+        "--model-name",
+        required=True,
+        help="HuggingFace model name compatible with AutoModelForSeq2SeqLM",
+    )
+    parser.add_argument(
+        "--state-dict-path",
+        required=True,
+        help="File path to the model's state_dict, can be in Cloud Storage",
+    )
+    args, beam_args = parser.parse_known_args()
+
+    logging.getLogger().setLevel(logging.INFO)
+    beam_options = PipelineOptions(
+        beam_args,
+        save_main_session=True,
+        pickle_library="cloudpickle",
```

Comment:
does the example work without this option? noting that it is still experimental. also, saving main session is a noop with cloudpickle, so you could remove that.

Reply:
Cloud pickle was needed to pickle the pipeline's PTransform, I wasn't aware it already pickled the main session so I'll remove it.